### PR TITLE
fix: assets are not encrypted at rest WPB-6977

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6977" title="WPB-6977" target="_blank">BugWPB-6977</a>  [iOS] Assets not encrypted at rest
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR adds support for encrypting assets at rest. Assets are stored on the file system in a cache directory. For any given message there may be one or more assets (e.g image data, raw file data, video file + thumbnail image), each of which may or may not be preprocessed (e.g images are downscaled) and encrypted for transport. Up until now, we only encrypted assets for transport then deleted them once the asset was uploaded. On the receiver side, we downloaded the encrypted asset, then decrypted it and stored only the decrypted asset data.

With this PR, we move to only storing the encrypted asset data and decrypt it on demand when accessing it. This means, for the sender, an asset will be stored in original unencrypted form, then the original is replaced by the preprocessed asset (if needed), and then the unencrypted data is replaced by the encrypted data. This is then uploaded to the backend. 

For the receiver side, we simply store the encrypted asset that was downloaded from the backend.

#### Test Coverage

- Updated unit tests

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Source/Model/Message/AssetEncryption.swift
Comment (caldrian) [inline wire-ios-data-model/Source/Model/Message/AssetEncryption.swift:1]: The file could be renamed to `FileAssetCacheExtensions` or something more specific.

File: wire-ios-data-model/Source/Model/Message/AssetEncryption.swift [hunk 1]
```diff
@@ -19,150 +19,313 @@
...
+        return encryptFileAndComputeSHA256Digest(
+            plaintextCacheKey,
+            encryptedEntryKey: encryptedCacheKey
+        )
+    }
+
+    // MARK: - Conversation message
+
+    /// Decrypts an encrypted asset in the asset cache to a decrypted version in the cache. 
+    /// Upon completion of the decryption, deletes the encrypted original. In case of error
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/AssetEncryption.swift]: > /// Upon completion of the decryption, deletes the encrypted original

question: why don't we keep the encrypted original

File: wire-ios-data-model/Source/Model/Conversation/Team.swift [hunk 2]
```diff
@@ -108,38 +108,55 @@ extension Team {
...
+                    changedKeys: [#keyPath(Team.imageData)],
+                    uiContext: uiContext
+                )
+            }
         }
     }
 
     public func requestImage() {
-        guard let moc = self.managedObjectContext, moc.zm_isUserInterfaceContext, !moc.zm_fileAssetCache.hasDataOnDisk(for: self, format: Team.defaultLogoFormat, encrypted: false) else { return }
+        guard
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/Conversation/Team.swift]: swiftlint

File: wire-ios-data-model/Source/Model/Message/V2Asset.swift [hunk 3]
```diff
@@ -75,10 +143,15 @@ public class V2Asset: NSObject, ZMImageMessageData {
         return FileAssetCache.cacheKeyForAsset(assetClientMessage, format: .preview)
     }
 
+    // TODO: [John] Check if we can delete this.
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/V2Asset.swift]: swiftlint  warning to suppress warning or resolve

File: wire-ios-data-model/Source/Model/Message/V3Asset.swift [hunk 4]
```diff
@@ -26,18 +26,15 @@ private let zmLog = ZMSLog(tag: "AssetV3")
 /// It only includes methods in which these two versions differentiate.
 @objc public protocol AssetProxyType {
 
+    // TODO: rename to "hasLocalFileData"
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/V3Asset.swift]: swiftlint  warning to suppress warning or resolve

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+LinkPreview.swift [hunk 5]
```diff
@@ -155,50 +154,88 @@ extension ZMClientMessage: ZMImageOwner {
...
                 $0.linkPreview = [linkPreview]
             }
 
             let messageUpdate: MessageCapable
+
             guard
                 let content = genericMessage.content,
-                let nonce = nonce else {
-                    return
+                let nonce = nonce
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+LinkPreview.swift]: swiftlint  warning

File: wire-ios-data-model/Source/Utilis/EAR/EAREncryptor.swift [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+public enum EAREncryptorError: Error {
+
+    case missingKey
+    case missingSalt
+    case failedToEncrypt(Error)
+    case failedToDecrypt(Error)
+
+}
+
+// TODO: doc comments
```
Comment (netbe) [inline wire-ios-data-model/Source/Utilis/EAR/EAREncryptor.swift]: swiftlint  warning to suppress warning or resolve

File: wire-ios-data-model/Source/Utilis/EAR/EAREncryptor.swift [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+public class EAREncryptor: EAREncryptorProtocol {
+
+    var key: VolatileData?
+    var salt: Data?
+
+    public init() {
+
```
Comment (netbe) [inline wire-ios-data-model/Source/Utilis/EAR/EAREncryptor.swift]: swiftlint  warning

File: wire-ios-request-strategy/Sources/Request Strategies/Assets/V2/ImageV2DownloadRequestStrategy.swift [hunk 8]
```diff
@@ -72,10 +72,20 @@ public final class ImageV2DownloadRequestStrategy: AbstractRequestStrategy {
...
 extension ImageV2DownloadRequestStrategy: ZMDownstreamTranscoder {
 
-    public func request(forFetching object: ZMManagedObject!, downstreamSync: ZMObjectSync!, apiVersion: APIVersion) -> ZMTransportRequest? {
-        guard let message = object as? ZMAssetClientMessage, let conversation = message.conversation else { return nil }
+    public func request(
+        forFetching object: ZMManagedObject!,
+        downstreamSync: ZMObjectSync!,
+        apiVersion: APIVersion
+    ) -> ZMTransportRequest? {
+        guard
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/Assets/V2/ImageV2DownloadRequestStrategy.swift]: swiftlint  warning

**Review by netbe — COMMENTED** [review]
just to resolve or suppress warnings from swiftlint

**Review by netbe — APPROVED** [review]

**Review by markusfassbender — APPROVED** [review]

**Pull request merged**